R/bainsembayesian.R

```diff
@@ -0,0 +1,150 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+BainSemBayesian <- function(jaspResults, dataset, options, ...) {
+
+  # What type of Bain analysis is being conducted?
+  type <- "sem"
+
+  # Read the data set
+  dataList <- .bainReadDataset(options, type, dataset)
+
+  # Check if current options allow for analysis
+  ready <- .bainOptionsReady(options, type, dataList[["dataset"]])
+
+  # Check if current data allow for analysis
+  .bainDataReady(dataList[["dataset"]], options, type)
+  
+  # Create a container for the results
+  bainContainer <- .bainGetContainer(jaspResults, deps = c("syntax", "model", "seed", "fraction", "standardized"))
+  
+  # Create a legend containing the order constrained hypotheses
+  .bainLegend(dataList[["dataset"]], options, type, jaspResults, position = 0)
+  
+  # Create a table containing the main analysis results
+  .bainResultsTable(dataList[["dataset"]], options, bainContainer, dataList[["missing"]], ready, type, position = 1)
+  
+  # Create the Bayes factor matrix
+  .bainBfMatrix(dataList[["dataset"]], options, bainContainer, ready, type, position = 2)
+  
+  # Create the descriptive statistics (coefficients) table
+  .bainDescriptivesTable(dataList[["dataset"]], options, bainContainer, ready, type, position = 3)
+  
+  # Create the posterior probability plots
+  .bainPosteriorProbabilityPlot(dataList[["dataset"]], options, bainContainer, ready, type, position = 4)
+  
+  # Create the path diagram plot
+  .bainSemPathDiagram(dataList[["dataset"]], options, bainContainer, ready, jaspResults, position = 5)
+}
+
+.bainLavaanState <- function(dataset, options, bainContainer, ready, jaspResults){
+  if(!is.null(bainContainer[["lavaanResult"]])){
+    return(bainContainer[["lavaanResult"]]$object)
+  } else if(ready){
+    syntax <- .bainSemTranslateModel(options[["syntax"]], dataset)
+    error <- try({
+      fit <- lavaan::sem(model = syntax, data = dataset)
+    })
+    if(isTryError(error)){
+      bainContainer$setError(gettextf("An error occurred in the call to lavaan: %1$s", jaspBase:::.extractErrorMessage(error)))
+      return()
+    }
+    bainContainer[["lavaanResult"]] <- createJaspState(fit)
+    bainContainer[["lavaanResult"]]$dependOn(options = "syntax")
+    return(bainContainer[["lavaanResult"]]$object)
+  }
+}
+
+.bainSemGetUsedVars <- function(syntax, availablevars, decode = FALSE) {
+  if(decode){
+    vv <- decodeColNames(availablevars)
+  } else {
+    vv <- availablevars
+  }
+  findpattern <- paste0("(?<=[\\s\\+\\^\\=\\~\\<\\*\\>\\:\\%\\|\\+]|^)\\Q",
+                        vv,
+                        "\\E(?=[\\s\\+\\^\\=\\~\\<\\*\\>\\:\\%\\|\\+]|$)")
```

Comment:
It's nice to put a little comment next to longer regex to show (an example of) what it's used for